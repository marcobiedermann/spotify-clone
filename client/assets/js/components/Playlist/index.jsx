import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import styles from './style.css';

const Playlist = ({
  id,
  images,
  name,
  owner,
}) => (
  <figure className={styles.playlist}>
    {images[0] && (
      <Link to={`/users/${owner.id}/playlists/${id}`}>
        <img src={images[0].url} alt={name} />
      </Link>
    )}
    <figcaption>
      <Link to={`/users/${owner.id}/playlists/${id}`}>
        {name}
      </Link>
    </figcaption>
  </figure>
);

Playlist.propTypes = {
  id: PropTypes.string,
  images: PropTypes.arrayOf(PropTypes.shape()),
  name: PropTypes.string,
  owner: PropTypes.shape({
    id: PropTypes.string,
  }),
};

Playlist.defaultProps = {
  id: '',
  images: [],
  name: '',
  owner: {
    id: '',
  },
};

export default Playlist;
