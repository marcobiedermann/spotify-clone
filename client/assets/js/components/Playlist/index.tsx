/* eslint-disable camelcase */

import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import Image, { ImageProps } from '../Image';
import { TracksProps } from '../Tracks';

interface OwnerProps {
  display_name: string;
  id: string;
}

export interface PlaylistProps {
  id: string;
  images: ImageProps[];
  name: string;
  owner: OwnerProps;
  tracks: TracksProps;
}

const Playlist: FC<PlaylistProps> = (props) => {
  const { id, images, name } = props;

  return (
    <figure>
      {images[0] && (
        <Link to={`/playlists/${id}`}>
          <Image {...images[0]} alt={name} />
        </Link>
      )}
      <figcaption>
        <Link to={`/playlists/${id}`}>{name}</Link>
      </figcaption>
    </figure>
  );
};

export default Playlist;
