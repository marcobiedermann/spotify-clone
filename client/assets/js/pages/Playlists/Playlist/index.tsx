import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { Link, RouteChildrenProps, useParams } from 'react-router-dom';
import useSWR from 'swr';
import Button from '../../../components/Button';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';
import Track from '../../../components/Track';

interface Params {
  playlistId: string;
}

const PlaylistPage: FC<RouteChildrenProps> = () => {
  const { playlistId } = useParams<Params>();
  const { data, error } = useSWR(`/v1/playlists/${playlistId}`);

  if (error) {
    return <Error>{error.message}</Error>;
  }

  if (!data) {
    return <Loader />;
  }

  const { name, owner, tracks } = data;

  return (
    <>
      <Helmet>
        <title>Playlist</title>
      </Helmet>
      <div>
        <p>Playlist</p>
        <h1>{name}</h1>
        <p>
          Created by <Link to={`/users/${owner.id}`}>{owner.display_name}</Link> Â· {tracks.total}{' '}
          songs
        </p>
        <p>
          <Button>Play</Button>
        </p>
        <table>
          <tbody>
            {tracks.items.map((track) => (
              <Track key={track.track.id} {...track.track} />
            ))}
          </tbody>
        </table>
      </div>
    </>
  );
};

export default PlaylistPage;
