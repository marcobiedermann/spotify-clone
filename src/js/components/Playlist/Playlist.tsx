import React from 'react';
import { Link } from 'react-router-dom';
import Image from '../Image';

interface Image {
  height: number;
  url: string;
  width: number;
}

export interface PlaylistProps {
  id: string;
  images: Image[];
  name: string;
}

function Playlist(props: PlaylistProps): JSX.Element {
  const { id, images, name } = props;

  return (
    <figure>
      {images[0] && (
        <Link to={`/playlists/${id}`}>
          <Image {...images[0]} alt={name} />
        </Link>
      )}
      <figcaption>
        <Link to={`/playlists/${id}`}>{name}</Link>
      </figcaption>
    </figure>
  );
}

export default Playlist;
