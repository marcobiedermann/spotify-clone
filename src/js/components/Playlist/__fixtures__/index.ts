const image = {
  height: 100,
  url: '',
  width: 100,
};

const playlist = {
  id: '',
  images: [image],
  name: '',
};

export default playlist;
