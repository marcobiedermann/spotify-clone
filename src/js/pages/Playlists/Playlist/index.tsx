import { formatDistance } from 'date-fns';
import React from 'react';
import { Helmet } from 'react-helmet';
import { Link, useParams } from 'react-router-dom';
import {
  Button,
  Error,
  Image,
  Loader,
  Media,
  MediaBody,
  MediaObject,
  PlaylistTracks,
  Section,
} from '../../../components';
import { usePlaylist } from '../../../hooks/playlists';

const defaultImageSize = 240;

function PlaylistPage(): JSX.Element {
  const { playlistId } = useParams();
  const { data, error, isError, isPending } = usePlaylist(playlistId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isPending) {
    return <Loader />;
  }

  const { description, images = [], name, owner, tracks } = data;
  const image = images[0];

  const totalDuration =
    tracks?.items.reduce(
      (accumulator, currentValue) => accumulator + (currentValue.track?.duration_ms || 0),
      0,
    ) || 0;

  return (
    <>
      <Helmet>
        <title>Playlist</title>
      </Helmet>
      <div>
        <header>
          <Media>
            <MediaObject>
              <Image
                {...image}
                alt={name || ''}
                width={defaultImageSize}
                height={defaultImageSize}
              />
            </MediaObject>
            <MediaBody>
              <h3>Playlist</h3>
              <h1>{name}</h1>
              {description && <p>{description}</p>}
              <p>
                <strong>
                  <Link to={`/users/${owner?.id}`}>{owner?.display_name}</Link> Â· {tracks?.total}{' '}
                  songs,
                </strong>{' '}
                {formatDistance(0, totalDuration)}
              </p>
            </MediaBody>
          </Media>
        </header>
        <Section>
          <p>
            <Button>Play</Button>
          </p>
        </Section>
        {tracks && (
          <Section>
            <PlaylistTracks items={tracks.items} />
          </Section>
        )}
      </div>
    </>
  );
}

export default PlaylistPage;
