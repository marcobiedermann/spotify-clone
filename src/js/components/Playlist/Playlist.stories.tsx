import type { Meta, StoryObj } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import Playlist from '.';
import playlistFixtures from './__fixtures__';

const meta: Meta<typeof Playlist> = {
  component: Playlist,
  args: playlistFixtures,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
};

type Story = StoryObj<typeof Playlist>;

const Default: Story = {};

export { Default };
export default meta;
