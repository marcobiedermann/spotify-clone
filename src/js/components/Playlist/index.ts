import Playlist from './Playlist';

export default Playlist;
export * from './Playlist';
