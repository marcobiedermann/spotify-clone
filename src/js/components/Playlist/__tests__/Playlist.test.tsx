import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { expect, it } from 'vitest';
import Playlist from '..';
import playlistFixtures from '../__fixtures__';

it('renders correctly', () => {
  const { container } = render(
    <MemoryRouter>
      <Playlist {...playlistFixtures} />
    </MemoryRouter>,
  );

  expect(container.firstChild).toMatchSnapshot();
});
