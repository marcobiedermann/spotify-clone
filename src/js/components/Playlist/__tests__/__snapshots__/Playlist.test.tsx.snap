// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<figure>
  <a
    href="/playlists/"
  >
    <img
      alt=""
      height="100"
      src=""
      width="100"
    />
  </a>
  <figcaption>
    <a
      href="/playlists/"
    />
  </figcaption>
</figure>
`;
